developing-without-a-build/article.md

```diff
@@ -0,0 +1,257 @@
+
+### HTML
+Over time web frameworks have solved html templating in different ways, for example by placing HTML inside javascript strings. Lately JSX has become a popular format for embedding HTML inside javascript, but it is not going to run natively in the browser without some kind of transformation.
+
+With es2015/es6 we can use tagged template string literals to embed html inside js, and use it for doing efficient DOM updates. This runs natively in the browser and still offers a great developer experience. There are some really good production ready and feature complete libraries that can be used for this:
+
+- [htm, JSX using template literals. Works with libraries that use JSX, such as react](https://www.npmjs.com/package/htm)
+- [lit-html, a html templating library](https://www.npmjs.com/package/lit-html)
+- [lit-element, integrates lit-html with web components](https://www.npmjs.com/package/lit-element)
+- [haunted, a functional web components library with react-like hooks](https://www.npmjs.com/package/haunted)
+- [hybrids, another functional web component library](https://www.npmjs.com/package/hybrids)
+- [hyperHTML, a html templating library](https://www.npmjs.com/package/hyperhtml)
+(TODO: more libraries?)
+
+For syntax highlighting you might need to configure your IDE or install a plugin.
+
+### CSS
+For HTML and JSON there are sufficient alternatives. Unfortunately with CSS it is more complicated. By itself CSS is not modular as it affects the entire page. A common complaint that this is what makes CSS so difficult to scale.
+
+There a lot of different ways to write css, it's beyond the scope of this article to look into all of them. When working without a build, regular stylesheets will just work fine of course.
+
+If you are using some kind of CSS preprocessor you can run it before running your web server and just load the CSS output. Many CSS in JS solutions should also work if the library publishes an es module format.
+
+#### Shadow dom
+For truly modular CSS I recommend looking into [Shadow dom](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM), it fixes many of the scoping and encapsulation problems of CSS. Unfortunately, it's not yet a complete story. There are still missing features that are being worked out in the standard so it may not yet be the right solution in all scenarios. (TODO: Link to article?)
+
+Good to mention here is [the lit-element library](https://www.npmjs.com/package/lit-element), which offers a great developer experience when authoring modular CSS without a build step. `lit-element` does most of the heavy lifting for you. You author css using tagged template literals, which are just syntax sugar for creating a [Constructable Stylesheet](https://developers.google.com/web/updates/2019/02/constructable-stylesheets). This way you can write and share CSS between your components. This will integrate well with css modules when they are shipped.
+
```

Comment:
is it possible to import css files as text, construct a stylesheet from them, and append to the shadow root?

Reply:
You mean today or when css modules ship?

Reply:
You can use fetch to grab a css files but its hard to optimize for production

Reply:
Added a section on this.